SCIENCE-503 `gcsfs` upstream update

VOvchinnikov opened the pull request:
https://remerge.atlassian.net/browse/SCIENCE-503

Metamess committed:
Add seek(0) to request data to prevent issues on retries (#624)

martindurant committed:
changelog (#626)

martindurant committed:
changelog (#629)

keunhong committed:
Handle case when error message is a string. (#631)

Fixes https://github.com/fsspec/gcsfs/issues/630.

martindurant committed:
changelog (#639)

* changelog

* use mincondax

* tr exact version

* more

* boost time

* debug

* auto creds

damjad committed:
Fix md5 hash logic. Make it order agnostic. (#640)


Co-authored-by: Martin Durant <martindurant@users.noreply.github.com>

Flinz committed:
Update credentials.py to test removal of race condition in credentials (#643)

martindurant committed:
changelog (#644)

amyewang committed:
Handle cache_timeout = 0 in GCSFileSystem (#646)

* add is not none

* test

rickmcgeer committed:
Added note about refresh and POSIX behavior (#648)



Co-authored-by: Martin Durant <martindurant@users.noreply.github.com>

martindurant committed:
avoid index error on whole second (#649)

martindurant committed:
Add exclusive write (#651)

martindurant committed:
changelog (#652)

kujenga committed:
Guess file content type when not specified (#655)



Co-authored-by: Martin Durant <martin.durant@alumni.utoronto.ca>

martindurant committed:
Try better shutdown cleanup behaviour (#657)

martindurant committed:
changelog (#660)

kujenga committed:
Improvements for credentials refresh under high load (#658)

martindurant committed:
changelog (#662)

martindurant committed:
Don't check expiry for creds that have none (#665)

martindurant committed:
Remove caveat from doc (#666)

martindurant committed:
Fix find with path not ending with "/" (#668)

martindurant committed:
changelog (#670)

martindurant committed:
Changelog (#671)

GuidoBallabio committed:
add userProject to batch deletion query (#673)

Bjoern-Rapp committed:
Yet another fix for isdir (#676)

kaiju committed:
Add support for specifying Cloud KMS keys when creating files (#679)


Co-authored-by: Martin Durant <martin.durant@alumni.utoronto.ca>

jlopezpena committed:
Avoid deprecated `utcnow` (#680)



---------

Co-authored-by: Martin Durant <martin.durant@alumni.utoronto.ca>

tassilucas committed:
Create warning for appending mode 'a' operations (#675)


---------

Co-authored-by: Martin Durant <martindurant@users.noreply.github.com>
Co-authored-by: Martin Durant <martin.durant@alumni.utoronto.ca>

martindurant committed:
changelog (#681)

martindurant committed:
Fix timezone comparison (#683)

stefannica committed:
Use correct timezone conversion for credentials expiry timestamp (#688)

Fixes #687

martindurant committed:
changelog (#689)

martindurant committed:
less refresh in ci (#690)

* less refresh in ci

* reinstate 401 retry (for now)

* add pytest-asyncio to test env

martindurant committed:
remove addcidental inclusion

martindurant committed:
add support message (#691)

martindurant committed:
changelog (#693)

+ rst->md for readme

martindurant committed:
Ensure right error type for get() on nonexistent (#695)

martindurant committed:
Fix iterating over every entry (#697)

martindurant committed:
changelog (#698)

martindurant committed:
Add goatcounter (#700)

martindurant committed:
Should not attempt to create bucket by default on mkdir (#701)

martindurant committed:
Add CoC (#703)

remerge-hal commented:
This comment ensures that the correct Slack channel is notified after the team/project label `SCIENCE` has been added to this pull request.

See [this comment](https://remerge.atlassian.net/browse/ITSUP-1973?focusedCommentId=90701) for details.

VOvchinnikov committed:
Restored "ghost" objects fix

VOvchinnikov committed:
Removed `test_dir_marker_...` tests

These tests are reliant on "ghost" objects (dir markers, I suppose) to be returned - and that's exactly what we _don't_ want to have.

VOvchinnikov committed:
Pinned `setup-miniconda` version

The reason being, there were attacks performed, substituting tags with the ones with the same name, but malicious code inside.

VOvchinnikov committed:
Merge branch 'master' into SCIENCE-503-gcsfs-upstream-update

VOvchinnikov committed:
Let `_info` get the listing _with_ markers intact
